modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Enrich.scala

```diff
@@ -0,0 +1,147 @@
+
+object Enrich {
+
+  /** Default adapter registry, can be constructed dynamically in future */
+  val adapterRegistry = new AdapterRegistry()
+
+  val processor: Processor = Processor(generated.BuildInfo.name, generated.BuildInfo.version)
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  /**
+   * Run a primary enrichment stream, reading from [[Environment]] source, enriching
+   * via [[enrichWith]] and sinking into [[GoodSink]] and [[BadSink]] respectively.
+   * Can be stopped via _stop signal_ from [[Environment]]
+   *
+   * The stream won't download any enrichment DBs. [[Assets]] stream or
+   * [[Assets.download]] should be responsible for that
+   */
+  def run[F[_]: Concurrent: ContextShift: Clock](env: Environment[F]): Stream[F, Unit] = {
+    val registry: F[EnrichmentRegistry[F]] = env.enrichments.get.map(_.registry)
+    val enrich: Enrich[F] = enrichWith[F](registry, env.igluClient, env.sentry, env.metrics.enrichLatency)
+    val badSink: BadSink[F] = _.evalTap(_ => env.metrics.badCount).through(env.bad)
+    val goodSink: GoodSink[F] = _.evalTap(_ => env.metrics.goodCount).through(env.good)
+
+    env.source
+      .pauseWhen(env.pauseEnrich)
+      .evalTap(_ => env.metrics.rawCount)
+      .parEvalMapUnordered(16)(payload => env.blocker.blockOn(enrich(payload)))
+      .flatMap(_.decompose[BadRow, EnrichedEvent])
+      .observeEither(badSink, goodSink)
+      .void
+  }
+
+  /**
+   * Enrich a single [[CollectorPayload]] to get list of bad rows and/or enriched events
+   *
+   * Along with actual `ack` the `enrichLatency` gauge will be updated
+   */
+  def enrichWith[F[_]: Clock: Sync](
+    enrichRegistry: F[EnrichmentRegistry[F]],
+    igluClient: Client[F, Json],
+    sentry: Option[SentryClient],
+    enrichLatency: Option[Long] => F[Unit]
+  )(
+    row: Payload[F, Array[Byte]]
+  ): F[Result[F]] = {
+    val payload = ThriftLoader.toCollectorPayload(row.data, processor)
+    val collectorTstamp = payload.toOption.flatMap(_.flatMap(_.context.timestamp).map(_.getMillis))
+
+    val result =
+      for {
+        _ <- Logger[F].debug(payloadToString(payload))
+        etlTstamp <- Clock[F].realTime(TimeUnit.MILLISECONDS).map(millis => new DateTime(millis))
+        registry <- enrichRegistry
+        enriched <- EtlPipeline.processEvents[F](adapterRegistry, registry, igluClient, processor, etlTstamp, payload)
+        trackLatency = enrichLatency(collectorTstamp)
+      } yield Payload(enriched, trackLatency *> row.finalise)
+
+    result.handleErrorWith(sendToSentry[F](row, sentry))
+  }
+
+  /** Stringify `ThriftLoader` result for debugging purposes */
+  def payloadToString(payload: ValidatedNel[BadRow.CPFormatViolation, Option[CollectorPayload]]): String =
+    payload.fold(_.asJson.noSpaces, _.map(_.toBadRowPayload.asJson.noSpaces).getOrElse("None"))
+
+  /** Transform enriched event into canonical TSV */
+  def encodeEvent(enrichedEvent: EnrichedEvent): String =
```

Comment:
Agreed. I actually think it should be a part of standard `Event` definition with ton of property-based tests: https://github.com/snowplow/snowplow-scala-analytics-sdk/issues/97

Reply:
Just in case - I wrote a small benchmarking module (https://github.com/snowplow/enrich/issues/370), which should make it easy to confirm/refute performance benefiits.